docs/raster/S3Streaming.md

```diff
@@ -0,0 +1,118 @@
+- - -
+### How to Use this Feature
+
+In order for streaming from S3 to work, two new types have been created. These are: ```StreamBytes``` and ```Bytereader```, respectively. the first deals with getting bytes in the form of ```Map[Long, Array[Byte]``` called a, "chunk". The ```Long``` of this chunk is the position in the file where the ```Array[Byte]``` starts. For example, if a new chunk needs to be created at the 100th byte of the file then the chunk that is created will have a value of 100. These chunks can vary in size and a new one is produced each time a call is made to another part of the file which is not contained within the current chunk.
+
+```ByteReader``` is what reads the bytes from the chunks. As of right now, only ```java.nio.ByteBuffer```s and the ```StreamByteReader``` type can access the chunk.
+
+#### Creating ```StreamBytes```
+There is only one way to construct an instance of ```StreamBytes``` as of this writing, and that is through the ```S3StreamBytes``` class. Found in ```geotrellis.spark.io.s3.util``` package, this class will be able to get chunks from a desired GeoTifff on S3. Bellow is a basic guide on creating this class:
+
+```scala
+ import geotrellis.spark.io.s3._
+ import geotrellis.spark.io.s3.util.S3StreamBytes
+ 
+ import com.amazonaws.services.s3.model._
+ 
+ val client: S3Client = S3Client.default
+ val bucket: String = "my-bucket"
+ val key: String = "path/to/my/geotiff.tif"
+ 
+ // how many bytes that should be streamed at a time
+ val chunkSize: Int = 256000
+ 
+ val s3bytes = S3StreamBytes(bucket, key, client, chunkSize)
+```
+
+#### Creating a ```ByteReader```
+A ```java.nio.ByteBuffer``` can be used as a ```ByteReader```, but because this focus is on streaming GeoTiffs off S3, only ```StreamByteReader``` will be looked at in this guide. Below is a basic guide on creating this class:
+
+```scala
+ import geotrellis.util.StreamByteReader
+ 
+ val reader = StreamByteReader(s3Bytes)
+```
+
+#### Streaming a GeoTiff From S3
+Now that we have a ```ByteReader``` all that needs to be done now is to supply it to the ```GeoTiffReader``` with any additional reading parameters that might be needed.
+
+#### Streaming the Entire GeoTiff
+Because GeoTiffs can be read in lazily now, files of any size should be able to be read in fully either from S3 or locally. However, precautions should be taken when doing so. It appears that when a file's size passes a certain point, the time it takes to perform operations on it degrades rapidly. This aspect of whole file streaming has not yet been looked into closely, so it is uncertain at what point performance begins to fall and to what extent. Therefore, it is best to use caution when streaming the entire file.
```

Comment:
I figured out how to do reading without using skip for the actual S3Client. Here's what it is [now](https://github.com/jbouffard/geotrellis/blob/wgt-s3/s3/src/main/scala/geotrellis/spark/io/s3/S3Client.scala#L173). I think it might have to do with the sheer size of the file and how many bytes it has to read through. Since to access the data the bytes need to be read in via a loop.